tools/torchci/clickhouse_query_perf.py

```diff
@@ -0,0 +1,185 @@
+"""
+Check query results and performance. Note that query performance is not stable
+and can vary significantly between runs.
+"""
+
+import argparse
+import json
+import subprocess
+import time
+from datetime import datetime, timedelta, timezone
+from functools import cache
+from typing import Optional
+
+from prettytable import PrettyTable
+from torchci.clickhouse import get_clickhouse_client, query_clickhouse
+from torchci.utils import REPO_ROOT
+from tqdm import tqdm
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description="Queue alert for torchci")
+    parser.add_argument("--query", type=str, help="Query name", required=True)
+    parser.add_argument(
+        "--perf", action="store_true", help="Run performance comparison"
+    )
+    parser.add_argument("--results", action="store_true", help="Run results comparison")
+    parser.add_argument(
+        "--times",
+        type=int,
+        help="Number of times to run the query. Only relevant if --perf is used",
+        default=10,
+    )
+    parser.add_argument(
+        "--compare",
+        type=str,
+        help="Either a sha or a branch name to compare against. These should be available locally. Required for --results",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def get_query_id(query: str, params: dict) -> Optional[str]:
+    try:
+        res = get_clickhouse_client().query(query, params)
+        return res.query_id
+    except Exception as e:
+        print(f"Error: {e}")
+        return None
+
+
+@cache
+def get_base_query(query: str, sha: str) -> str:
+    return subprocess.check_output(
+        ["git", "show", f"{sha}:torchci/clickhouse_queries/{query}/query.sql"]
+    ).decode("utf-8")
+
+
+EXECUTION_METRICS = """
+SELECT
+    round(avg(query_duration_ms)) AS realTimeMSAvg,
+    avg(memory_usage) as memoryBytesAvg
+FROM
+    clusterAllReplicas(default, system.query_log)
+where
+    has({query_ids: Array(String)}, query_id)
+    and type = 'QueryFinish'
+"""
+
+
+def get_avg_stats(query_ids: list) -> tuple:
+    metrics = query_clickhouse(EXECUTION_METRICS, {"query_ids": query_ids})
+    return metrics[0]["realTimeMSAvg"], metrics[0]["memoryBytesAvg"]
+
+
+def get_query_ids(query: str, params: dict, times: int) -> tuple:
+    return [
+        x for _ in tqdm(range(times)) if (x := get_query_id(query, params)) is not None
+    ]
+
+
+def format_comparision_string(new: float, old: float) -> str:
+    return f"{new} vs {old} ({new - old}, {round(100 * (new - old) / old)}%)"
+
+
+@cache
+def get_query(query: str) -> tuple:
+    with open(
+        REPO_ROOT / "torchci" / "clickhouse_queries" / query / "params.json"
+    ) as f:
+        tests = json.load(f).get("tests", [])
+    with open(REPO_ROOT / "torchci" / "clickhouse_queries" / query / "query.sql") as f:
+        query = f.read()
+    for test in tests:
```

Comment:
Why isn't this part in `get_base_query` too?  I think we should be able to have just one `get_query` function that accepts the query name and the SHA, the use it to get both the current and the base queries.

Also the edge cases might be when adding a new query or, in rarer case, to delete one.

Reply:
Only downside is now you have to make a commit in order to run the script